### PR TITLE
[MLIR][Presburger] Template Matrix to allow MPInt and Fraction

### PR DESCRIPTION
**Pull request by Abhinav271828**
Matrix has been templated to Matrix<T> (for MPInt and Fraction) with explicit instantiation for both these types.
makeMatrix has been duplicated to makeIntMatrix and makeFracMatrix.

### CODE REVIEW

File: mlir/include/mlir/Analysis/Presburger/Matrix.h [hunk 1]
```diff
@@ -32,7 +34,10 @@ namespace presburger {
 /// (i, j) is stored at data[i*nReservedColumns + j]. The reserved but unused
 /// columns always have all zero values. The reserved rows are just reserved
 /// space in the underlying SmallVector's capacity.
+template<typename T>
 class Matrix {
+  // This class is not intended for general use: it supports only integers and rational numbers
```
Comment (Superty) [inline mlir/include/mlir/Analysis/Presburger/Matrix.h:39]: Please write full sentences in comments, with full stops

File: mlir/lib/Analysis/Presburger/Matrix.cpp [hunk 2]
```diff
@@ -375,3 +376,12 @@ bool Matrix::hasConsistentState() const {
 #endif
   return true;
 }
+
+namespace mlir
+{
+namespace presburger
+{
+template class Matrix<MPInt>;
+template class Matrix<Fraction>;
```
Comment (Superty) [inline mlir/lib/Analysis/Presburger/Matrix.cpp:385]: Please split the Fraction instance and related changes to another patch

**Review by superty — CHANGES_REQUESTED** [review]

File: mlir/include/mlir/Analysis/Presburger/Matrix.h [hunk 3]
```diff
@@ -32,7 +34,10 @@ namespace presburger {
 /// (i, j) is stored at data[i*nReservedColumns + j]. The reserved but unused
 /// columns always have all zero values. The reserved rows are just reserved
 /// space in the underlying SmallVector's capacity.
+template<typename T>
 class Matrix {
+  // This class is not intended for general use: it supports only integers and rational numbers.
```
Comment (Superty) [inline mlir/include/mlir/Analysis/Presburger/Matrix.h:39]: Can you move this comment two lines up above the template<typename

**Comment by NathanielMcVicar** [comment]
It looks like this has broken the Windows MLIR bot: https://lab.llvm.org/buildbot/#/builders/13/builds/40242/steps/6/logs/stdio
```
FAILED: tools/mlir/lib/Analysis/Presburger/CMakeFiles/obj.MLIRPresburger.dir/Matrix.cpp.obj 
C:\buildbot\mlir-x64-windows-ninja\llvm-project\mlir\lib\Analysis\Presburger\Matrix.cpp(385): error C2220: the following warning is treated as an error
C:\buildbot\mlir-x64-windows-ninja\llvm-project\mlir\lib\Analysis\Presburger\Matrix.cpp(385): warning C4661: 'T mlir::presburger::Matrix<T>::normalizeRow(unsigned int)': no suitable definition provided for explicit template instantiation request
        with
        [
            T=mlir::presburger::Fraction
        ]
C:\buildbot\mlir-x64-windows-ninja\llvm-project\mlir\include\mlir/Analysis/Presburger/Matrix.h(165): note: see declaration of 'mlir::presburger::Matrix<mlir::presburger::Fraction>::normalizeRow'
```

**Pull request merged**